### DISCUSSION TITLE
I want to detect two Apriltag with same id

### DISCUSSION THREAD

**Discussion by aliyoob** [FIRST-Tech-Challenge/FtcRobotController#673] [General]
Hi
I want to detect two Apriltag with same id
But there is a problem, the program can't distinguish the two tags. The numbers constantly keep changing and replacing each other. For example Yaw or distance are constantly keep changing between 6 to 14

**Comment by Windwoes** [comment]
My [reply](https://github.com/FIRST-Tech-Challenge/ftcdocs/issues/218#issuecomment-1708754003) to your post on the [ftc docs issue](https://github.com/FIRST-Tech-Challenge/ftcdocs/issues/218) still holds :)

**Reply by aliyoob** [reply]
Yeah thanks 
I solve that problem just now

**Comment by Shriharsh-Deshmukh** [comment]
To address this, you can consider the following approaches:

### 1. **Use Different IDs for Each Tag**
   If possible, assign different IDs to each AprilTag. This would allow the detection system to track each tag independently without confusion.

### 2. **Track Tags by Their Position**
   If you must use tags with the same ID, you can try to differentiate the tags by their positions relative to each other. For example, you can filter the detection results based on their spatial position (e.g., the distance between the two tags or their relative orientation). This would require tracking the tags' coordinates over time and determining which one is which based on their positions.

### 3. **Filter Based on Movement or Stabilization**
   If the tags are moving in different patterns, you could try stabilizing the measurements by filtering out sudden changes or applying a moving average filter to smooth out the readings. This could help mitigate the issue of rapidly changing values like yaw or distance.

### 4. **Use Multi-View Detection**
   If you can place cameras in different positions or use multi-camera systems, you may be able to detect the same tag from multiple angles and differentiate them based on where they are in the frame.

Let me know which approach fits your scenario best.

**Comment by gearsincorg** [comment]
I guess my question is "why do you need to detect two tags with the same ID".   It's like having two houses on a street with the same house number.  It kinda defeats the purpose of having house numbers.   

If we knew the reason, perhaps we could recommend an alternative solution.

**Reply by AlecHub** [reply]
Perhaps @aliyoob was unaware that you can (and should) train your vision device to assign unique IDs to the tags of interest.

**Reply by gearsincorg** [reply]
Because the poster is talking about the tag's Yaw and Distance, I assume they are refering to using the SDK's built-in AprilTag processing.   (Husky Lens or other AI based training don't provide yaw data).   So there is no opportunity to "train" the detection processs.  It simply  returns the actual ID of the specific apriltag being used.    The guideline here is that you should NOT normally use two tags with the same ID.  (The FTC field always uses unique tag ids)  Which is why I wanted to know why they needed to distinguish between them.

**Reply by aliyoob** [reply]
In the FIRST Global 2023 competition, we had a hydrogen tank with AprilTags on each of its four sides. We needed to detect two of these tags simultaneously to navigate from the corner towards the front and capture both sides. To do this, we needed to detect two tags with the same ID. I solved this by using an ‘if’ statement: if the "yaw" was positive and greater than zero, it would consider one tag for the right side and the other for the left. This way, I managed to solve the problem and find the correct position.

**Reply by gearsincorg** [reply]
Hi.  That was a challenging problem, and you found a great solution.   Having multiple identical Tag ID's on the field was probably not the best idea :)
